File: tests/bin/run-tests.py
#!/usr/bin/env python3
"""Simple test runner. No dependencies beyond parable itself."""

import os
import sys
import time


def find_test_files(directory):
    """Find all .tests files recursively. Returns list of paths."""
    result = []
    for root, _dirs, files in os.walk(directory):
        for f in files:
            if f.endswith(".tests"):
                result.append(os.path.join(root, f))
    result.sort()
    return result


def parse_test_file(filepath):
    """Parse a .tests file. Returns list of (name, input, expected, line_num) tuples."""
    tests = []
    with open(filepath) as f:
        lines = f.read().split("\n")
    i = 0
    n = len(lines)
    while i < n:
        line = lines[i]
        # Skip comments and blank lines
        if line.startswith("#") or line.strip() == "":
            i = i + 1
            continue
        # Test header
        if line.startswith("=== "):
            name = line[4:].strip()
            start_line = i + 1
            i = i + 1
            # Collect input until ---
            input_lines = []
            while i < n and lines[i] != "---":
                input_lines.append(lines[i])
                i = i + 1
            # Skip ---
            if i < n and lines[i] == "---":
                i = i + 1
            # Collect expected until ---, next test, or EOF
            expected_lines = []
            while i < n and lines[i] != "---" and not lines[i].startswith("=== "):
                expected_lines.append(lines[i])
                i = i + 1
            # Skip --- end marker
            if i < n and lines[i] == "---":
                i = i + 1
            # Strip trailing blank lines from expected
            while len(expected_lines) > 0 and expected_lines[-1].strip() == "":
                expected_lines.pop()
            test_input = "\n".join(input_lines)
            test_expected = "\n".join(expected_lines)
            tests.append((name, test_input, test_expected, start_line))
        else:
            i = i + 1
    return tests


def normalize(s):
    """Normalize whitespace for comparison."""
    return " ".join(s.split())


def run_test(test_input, test_expected):
    """Run a single test. Returns (passed, actual, error_msg)."""
    from parable import ParseError, parse

    try:
        nodes = parse(test_input)
        actual = " ".join(node.to_sexp() for node in nodes)
    except ParseError as e:
        return (False, "<parse error>", str(e))
    except Exception as e:
        return (False, "<exception>", str(e))

    expected_norm = normalize(test_expected)
    actual_norm = normalize(actual)
    if expected_norm == actual_norm:
        return (True, actual, None)
    else:
        return (False, actual, None)


def main():
    # Find test directory
    script_dir = os.path.dirname(os.path.abspath(__file__))
    test_dir = os.path.dirname(script_dir)  # tests/
    repo_root = os.path.dirname(test_dir)

    # Parse arguments
    verbose = False
    filter_pattern = None
    i = 1
    while i < len(sys.argv):
        arg = sys.argv[i]
        if arg == "-v" or arg == "--verbose":
            verbose = True
        elif arg == "-f" or arg == "--filter":
            i = i + 1
            if i < len(sys.argv):
                filter_pattern = sys.argv[i]
        else:
            # Assume it's a test directory or file
            if os.path.exists(arg):
                test_dir = arg
        i = i + 1

    # Find and run tests
    start_time = time.time()
    total_passed = 0
    total_failed = 0
    failed_tests = []

    if os.path.isfile(test_dir):
        test_files = [test_dir]
    else:
        test_files = find_test_files(test_dir)

    for filepath in test_files:
        tests = parse_test_file(filepath)
        rel_path = os.path.relpath(filepath, repo_root)

        for name, test_input, test_expected, line_num in tests:
            # Apply filter
            if filter_pattern is not None:
                if filter_pattern not in name and filter_pattern not in rel_path:
                    continue

            passed, actual, error_msg = run_test(test_input, test_expected)

            if passed:
                total_passed = total_passed + 1
                if verbose:
                    print(f"PASS {rel_path}:{line_num} {name}")
            else:
                total_failed = total_failed + 1
                failed_tests.append(
                    (rel_path, line_num, name, test_input, test_expected, actual, error_msg)
                )
                if verbose:
                    print(f"FAIL {rel_path}:{line_num} {name}")

    elapsed = time.time() - start_time

    # Print summary
    print("")
    if total_failed > 0:
        print("=" * 60)
        print("FAILURES")
        print("=" * 60)
        for rel_path, line_num, name, inp, expected, actual, error_msg in failed_tests:
            print(f"\n{rel_path}:{line_num} {name}")
            print(f"  Input:    {inp!r}")
            print(f"  Expected: {expected}")
            print(f"  Actual:   {actual}")
            if error_msg is not None:
                print(f"  Error:    {error_msg}")
        print("")

    print(f"{total_passed} passed, {total_failed} failed in {elapsed:.2f}s")

    if total_failed > 0:
        sys.exit(1)
    sys.exit(0)


if __name__ == "__main__":
    main()

File: tests/bin/run-tests.lua
#!/usr/bin/env lua
-- Simple test runner for Lua backend. Mirrors run-tests.py.

-- Find all .tests files recursively
local function find_test_files(directory)
  local result = {}
  local handle = io.popen('find "' .. directory .. '" -name "*.tests" -type f | sort')
  if handle then
    for line in handle:lines() do
      result[#result + 1] = line
    end
    handle:close()
  end
  return result
end

-- Read file contents
local function read_file(filepath)
  local f = io.open(filepath, "r")
  if not f then return nil end
  local content = f:read("*all")
  f:close()
  return content
end

-- Split string by newlines
local function split_lines(s)
  local lines = {}
  for line in (s .. "\n"):gmatch("([^\n]*)\n") do
    lines[#lines + 1] = line
  end
  return lines
end

-- Parse a .tests file. Returns array of {name, input, expected, line_num}
local function parse_test_file(filepath)
  local tests = {}
  local content = read_file(filepath)
  if not content then return tests end
  local lines = split_lines(content)
  local i = 1
  local n = #lines
  while i <= n do
    local line = lines[i]
    -- Skip comments and blank lines
    if line:match("^#") or line:match("^%s*$") then
      i = i + 1
    -- Test header
    elseif line:match("^=== ") then
      local name = line:sub(5):match("^%s*(.-)%s*$")
      local start_line = i
      i = i + 1
      -- Collect input until ---
      local input_lines = {}
      while i <= n and lines[i] ~= "---" do
        input_lines[#input_lines + 1] = lines[i]
        i = i + 1
      end
      -- Skip ---
      if i <= n and lines[i] == "---" then
        i = i + 1
      end
      -- Collect expected until ---, next test, or EOF
      local expected_lines = {}
      while i <= n and lines[i] ~= "---" and not lines[i]:match("^=== ") do
        expected_lines[#expected_lines + 1] = lines[i]
        i = i + 1
      end
      -- Skip --- end marker
      if i <= n and lines[i] == "---" then
        i = i + 1
      end
      -- Strip trailing blank lines from expected
      while #expected_lines > 0 and expected_lines[#expected_lines]:match("^%s*$") do
        expected_lines[#expected_lines] = nil
      end
      local test_input = table.concat(input_lines, "\n")
      local test_expected = table.concat(expected_lines, "\n")
      tests[#tests + 1] = {name = name, input = test_input, expected = test_expected, line_num = start_line}
    else
      i = i + 1
    end
  end
  return tests
end

-- Normalize whitespace for comparison
local function normalize(s)
  -- Replace invalid UTF-8 with ?
  s = s:gsub("[\128-\255]+", "?")
  -- Split on whitespace and rejoin with single spaces
  local parts = {}
  for word in s:gmatch("%S+") do
    parts[#parts + 1] = word
  end
  return table.concat(parts, " ")
end

-- Timeout via instruction counting (100k hook calls ~ plenty of time)
local HOOK_LIMIT = 100000

local function run_with_timeout(func)
  local count = 0
  local function hook()
    count = count + 1
    if count > HOOK_LIMIT then
      error("timeout")
    end
  end
  debug.sethook(hook, "", 100)  -- Hook every 100 VM instructions
  local ok, result = pcall(func)
  debug.sethook()  -- Clear the hook
  return ok, result
end

-- Run a single test. Returns passed, actual, error_msg
local function run_test(test_input, test_expected)
  -- Check for @extglob directive
  local extglob = false
  if test_input:match("^# @extglob\n") then
    extglob = true
    test_input = test_input:sub(#"# @extglob\n" + 1)
  end

  local ok, result = run_with_timeout(function()
    local nodes = parse(test_input, extglob)
    local sexps = {}
    for _, node in ipairs(nodes) do
      sexps[#sexps + 1] = node:to_sexp()
    end
    return table.concat(sexps, " ")
  end)

  if not ok then
    -- Parse error, timeout, or exception
    if result == "timeout" then
      if normalize(test_expected) == "<infinite>" then
        return true, "<infinite>", nil
      end
      return false, "<timeout>", "Test exceeded instruction limit"
    end
    if normalize(test_expected) == "<error>" then
      return true, "<error>", nil
    end
    return false, "<parse error>", tostring(result)
  end

  local actual = result

  -- If we expected an error but got a successful parse, that's a failure
  if normalize(test_expected) == "<error>" then
    return false, actual, "Expected parse error but got successful parse"
  end

  local expected_norm = normalize(test_expected)
  local actual_norm = normalize(actual)
  if expected_norm == actual_norm then
    return true, actual, nil
  else
    return false, actual, nil
  end
end

local function main()
  -- Parse arguments
  local verbose = false
  local filter_pattern = nil
  local test_dir = nil

  local i = 1
  while i <= #arg do
    local a = arg[i]
    if a == "-v" or a == "--verbose" then
      verbose = true
    elseif a == "-f" or a == "--filter" then
      i = i + 1
      if i <= #arg then
        filter_pattern = arg[i]
      end
    else
      test_dir = a
    end
    i = i + 1
  end

  if not test_dir then
    -- Get script directory
    local script_path = arg[0]
    local script_dir = script_path:match("(.*/)")
    if script_dir then
      test_dir = script_dir:gsub("/bin/$", "")
    else
      test_dir = "."
    end
  end

  -- Find and run tests
  local start_time = os.clock()
  local total_passed = 0
  local total_failed = 0
  local failed_tests = {}

  local test_files
  local attr = io.open(test_dir, "r")
  if attr then
    attr:close()
    -- Check if it's a file or directory
    local handle = io.popen('test -f "' .. test_dir .. '" && echo file || echo dir')
    if handle then
      local result = handle:read("*l")
      handle:close()
      if result == "file" then
        test_files = {test_dir}
      else
        test_files = find_test_files(test_dir)
      end
    else
      test_files = find_test_files(test_dir)
    end
  else
    test_files = find_test_files(test_dir)
  end

  for _, filepath in ipairs(test_files) do
    local tests = parse_test_file(filepath)
    local rel_path = filepath:gsub(".*/tests/", "tests/")

    for _, test in ipairs(tests) do
      local name = test.name
      local test_input = test.input
      local test_expected = test.expected
      local line_num = test.line_num

      -- Apply filter
      if filter_pattern then
        if not (name:find(filter_pattern, 1, true) or rel_path:find(filter_pattern, 1, true)) then
          goto continue
        end
      end

      -- Treat <infinite> as <error>
      local effective_expected = test_expected
      if normalize(test_expected) == "<infinite>" then
        effective_expected = "<error>"
      end

      -- Show which test is running (for debugging hangs)
      io.stderr:write(string.format("\r[%d] %s:%d %s                    ", total_passed + total_failed + 1, rel_path, line_num, name:sub(1,30)))
      io.stderr:flush()

      local passed, actual, error_msg = run_test(test_input, effective_expected)

      if passed then
        total_passed = total_passed + 1
        if verbose then
          print(string.format("PASS %s:%d %s", rel_path, line_num, name))
        end
      else
        total_failed = total_failed + 1
        failed_tests[#failed_tests + 1] = {
          rel_path = rel_path,
          line_num = line_num,
          name = name,
          input = test_input,
          expected = test_expected,
          actual = actual,
          error_msg = error_msg
        }
        if verbose then
          print(string.format("FAIL %s:%d %s", rel_path, line_num, name))
        end
      end

      ::continue::
    end
  end

  local elapsed = os.clock() - start_time
  io.stderr:write("\r" .. string.rep(" ", 40) .. "\r")  -- Clear progress line

  -- Print summary
  print("")
  if total_failed > 0 then
    print(string.rep("=", 60))
    print("FAILURES")
    print(string.rep("=", 60))
    for _, f in ipairs(failed_tests) do
      print(string.format("\n%s:%d %s", f.rel_path, f.line_num, f.name))
      print(string.format("  Input:    %q", f.input))
      print(string.format("  Expected: %s", f.expected))
      print(string.format("  Actual:   %s", f.actual))
      if f.error_msg then
        print(string.format("  Error:    %s", f.error_msg))
      end
    end
    print("")
  end

  print(string.format("%d passed, %d failed in %.2fs", total_passed, total_failed, elapsed))

  if total_failed > 0 then
    os.exit(1)
  else
    os.exit(0)
  end
end

main()
